lib/AST/ASTImporter.cpp

```diff
@@ -8154,15 +8154,22 @@ Expected<FileID> ASTImporter::Import(FileID FromID) {
       // than mmap the files several times.
       const FileEntry *Entry =
           ToFileManager.getFile(Cache->OrigEntry->getName());
-      if (!Entry)
-        // FIXME: Use a new error kind?
-        return llvm::make_error<ImportError>(ImportError::Unknown);
-      ToID = ToSM.createFileID(Entry, *ToIncludeLoc,
-                               FromSLoc.getFile().getFileCharacteristic());
-    } else {
+      // FIXME: The filename may be a virtual name that does probably not
+      // point to a valid file and we get no Entry here. In this case try with
+      // the memory buffer below.
+      if (Entry)
+        ToID = ToSM.createFileID(Entry, *ToIncludeLoc,
+                                 FromSLoc.getFile().getFileCharacteristic());
+    }
+    if (ToID.isInvalid()) {
       // FIXME: We want to re-use the existing MemoryBuffer!
+      bool Invalid = true;
       const llvm::MemoryBuffer *FromBuf =
-          Cache->getBuffer(FromContext.getDiagnostics(), FromSM);
+          Cache->getBuffer(FromContext.getDiagnostics(), FromSM, SourceLocation{}, &Invalid);
```

Comment:
Is it possible to have `FromBuf` as null and Invalid as `false`? It just does not make sense to have the `Invalid` variable.

Reply:
I do not know if `FromBuf` can be null at all (there was no test) or is it possible to get true `Invalid` but it is more sure to have check.

Reply:
ok

Reply:
Even from the source code of `getBuffer` it is not obvious that the returned value can not be null.